airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/PostgresSqlOperations.java

```diff
@@ -0,0 +1,84 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.destination.postgres;
+
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.db.jdbc.JdbcDatabase;
+import io.airbyte.integrations.destination.jdbc.JdbcSqlOperations;
+import io.airbyte.protocol.models.AirbyteRecordMessage;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.sql.SQLException;
+import java.util.List;
+import org.postgresql.copy.CopyManager;
+import org.postgresql.core.BaseConnection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PostgresSqlOperations extends JdbcSqlOperations {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PostgresSqlOperations.class);
+
+  @Override
+  public void insertRecords(JdbcDatabase database, List<AirbyteRecordMessage> records, String schemaName, String tmpTableName) throws SQLException {
+    if (records.isEmpty()) {
+      return;
+    }
+
+    database.execute(connection -> {
+      File tmpFile = null;
+      try {
+        tmpFile = Files.createTempFile(tmpTableName + "-", ".tmp").toFile();
+        writeBatchToFile(tmpFile, formatRecords(records));
+
+        var copyManager = new CopyManager(connection.unwrap(BaseConnection.class));
+        var sql = String.format("COPY %s.%s FROM stdin DELIMITER ',' CSV", schemaName, tmpTableName);
+        var bufferedReader = new BufferedReader(new FileReader(tmpFile));
+        copyManager.copyIn(sql, bufferedReader);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      } finally {
+        try {
+          if (tmpFile != null) {
+            Files.delete(tmpFile.toPath());
+          }
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    });
+  }
+
+  private List<AirbyteRecordMessage> formatRecords(List<AirbyteRecordMessage> records) {
+    // Postgres fails if json contains \u0000 unicode (NULL) in a json.
+    records.forEach(airbyteRecordMessage -> airbyteRecordMessage
+        .setData(Jsons.deserialize(Jsons.serialize(airbyteRecordMessage.getData()).replaceAll("\\\\u0000", ""))));
```

Comment:
This looks fine but my concern is that doing the `Jsons.deserialize(Jsons.serialize` for each record here is going to have a performance impact. How about we move this to `BufferedStreamConsumer`. We already have a string conversion [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java#L154) so it would save us from doing the serialization twice.

Reply:
Good point! Thanks ;)